[JSON manual] json_merge_preserve & json_merge_patch (#66)
CUBRID/cubrid-manual: en/sql/function/json_fn.rst

      '2'
      '3'
      NULL

JSON_MERGE_PATCH
===================================

.. function:: JSON_MERGE_PATCH (json_doc, json_doc [, json_doc] ...)

> The **JSON_MERGE_PATCH** function merges 2 or more json docs and returns the resulting merged json. **JSON_MERGE_PATCH** differs from **JSON_MERGE_PRESERVE** in that it will take the second argument when encountering merging conflicts. **JSON_MERGE_PATCH** is compliant with 

Reviewer: Add an empty line after `.. function:: JSON_MERGE_PATCH (json_doc, json_doc [, json_doc] ...)`

`RFC 7396 <https://tools.ietf.org/html/rfc7396/>`_.

> The merging of 2 json docuemnts is performed after the following rules, recursively:

Reviewer: ```suggestion
The merging of 2 json documents is performed with the following rules, recursively:
```

- when first argument is not an object, the result of the merge is the second object. As an exception, when the second argument is an object, the result of the merge is the merge result of the second argument with an empty object.
- when 2 objects are merged, the resulting object consists of the following (key, value) pairs:

  - All (key, value) pairs from the first object that have no corresponding (key, value) pairs in the second object.
  - All (key, value) pairs from the second object that have no corresponding (key, value) pairs in the first object, having values not null.

>   - All (key, value) pairs with same keys that exist in both objects, the second object's pair having the value not null. The values of these pairs become the results of the merging operations performed according to their json type merging rules.

Reviewer: Please rephrase third case when same key is found in both objects, current explanation is unclear. Something along these lines:
  `If second object value is null ... If second object value is ? ... If both values are json objects, apply rules recursively.`

> The JSON_MERGE_PATCH of k>=3 json documents is equivalent to JSON_MERGE_PATCH applied on first k-1 arguments and then applying a JSON_MERGE_PATCH on the result of the first JSON_MERGE_PATCH and the kth argument. 

Reviewer: ```suggestion
Merge operations are executed serially when there are more than two arguments: the result of merging first two arguments is merged with third, this result is then merged with fourth and so on.
```

::

      json_merge_patch('["a","b","c"]', '"scalar"')
    ======================
      "scalar"

The exception to the merge-patching, when the first argument is non-object and the second is an object. A merge operation is performed between an empty object and the second object argument.

.. code-block:: sql

    SELECT JSON_MERGE_PATCH ('["a"]', '{"a":null}');
::

      json_merge_patch('["a"]', '{"a":null}')
    ======================
      {}

Objects merging example, exemplifying the described object merging rules:

.. code-block:: sql

    SELECT JSON_MERGE_PATCH ('{"a":null,"c":["elem"]}','{"b":null,"c":{"k":null},"d":"elem"}');
::

      json_merge_patch('{"a":null,"c":["elem"]}', '{"b":null,"c":{"k":null},"d":"elem"}')
    ======================
      {"a":null,"c":{},"d":"elem"}

JSON_MERGE_PRESERVE
===================================

.. function:: JSON_MERGE_PRESERVE (json_doc, json_doc [, json_doc] ...)

  The **JSON_MERGE_PRESERVE** function merges 2 or more json docs and returns the resulting merged json. **JSON_MERGE_PRESERVE** differs from **JSON_MERGE_PATCH** in that it preserves both json elements on merging conflicts.

>   The merging of 2 json docuemnts is performed after the following rules, recursively:

Reviewer: ```suggestion
  The merging of 2 json documents is performed after the following rules, recursively:
```

- when 2 json arrays are merged, they are concatenated.

> - when 2 json objects are merged, all pairs are kept that do not have a corresponding pair in the other json object. The pairs, that have a corresponding pair in the other json object are combined in a pair with the same key and with the two values being merged according to their json type merging rules.

Reviewer: ```suggestion
- when 2 json objects are merged, all pairs that do not have a corresponding pair in the other json object are preserved. For matching keys, the values are always merged by applying the rules recursively.
```
It is better to let json object rules last.

> - when 2 non-array (scalar/object) json elements are merged and at most one of them is a json object, they are wrapped as singletons and merged as 2 json arrays.

Reviewer: ```suggestion
- when 2 non-array (scalar/object) json elements are merged and at most one of them is a json object, the result is an array containing the two json elements.
```

> - when a non-array json element is merged with a json array, the non-array is wrapped as a singleton json array and then merged with the json array according to json array merging rules.

Reviewer: ```suggestion
- when a non-array json element is merged with a json array, the non-array is wrapped as a single element json array and then merged with the json array according to json array merging rules.
```

>   The JSON_MERGE_PRESERVE operation on k>=3 json documents is equivalent to JSON_MERGE_PRESERVE applied on first k-1 arguments and then applying a JSON_MERGE_PRESERVE on the result of the first JSON_MERGE_PRESERVER and the kth argument.

Reviewer: ```suggestion
  Merge operations are executed serially when there are more than two arguments: the result of merging first two arguments is merged with third, this result is then merged with fourth and so on.
```

      json_contains('["a", "b1", ["a", {"k":["b2"]}]]','["b1", {"k":"b2"}]')
    ========================================================================
                                                                           1

JSON_MERGE_PATCH
===================================

.. function:: JSON_MERGE_PATCH (json_doc, json_doc [, json_doc] ...)

The **JSON_MERGE_PATCH** function merges 2 or more json docs and returns the resulting merged json. **JSON_MERGE_PATCH** differs from **JSON_MERGE_PRESERVE** in that it will take the second argument when encountering merging conflicts. **JSON_MERGE_PATCH** is compliant with
`RFC 7396 <https://tools.ietf.org/html/rfc7396/>`_.

The merging of 2 json documents is performed with the following rules, recursively:

> - when first argument is not an object, the result of the merge is the second object. As an exception, when the second argument is an object, the result of the merge is the merge result of the second argument with an empty object.

Reviewer: Might be easier to understand if this line is split into subcases - no argument is json object, first is a json object, second is not, first is not, second is.

      json_contains('["a", "b1", ["a", {"k":["b2"]}]]','["b1", {"k":"b2"}]')
    ========================================================================
                                                                           1

JSON_MERGE_PATCH
===================================

.. function:: JSON_MERGE_PATCH (json_doc, json_doc [, json_doc] ...)

> The **JSON_MERGE_PATCH** function merges 2 or more json docs and returns the resulting merged json. **JSON_MERGE_PATCH** differs from **JSON_MERGE_PRESERVE** in that it will take the second argument when encountering merging conflicts. **JSON_MERGE_PATCH** is compliant with

Reviewer 2: ```suggestion
The **JSON_MERGE_PATCH** function merges two or more json docs and returns the resulting merged json. **JSON_MERGE_PATCH** differs from **JSON_MERGE_PRESERVE** in that it will take the second argument when encountering merging conflicts. **JSON_MERGE_PATCH** is compliant with
```
